### PR TITLE
ytdl_hook: improve handling of json parsing errors

### PR DESCRIPTION
**Pull request by Retusthese**
ytdl_hook handles JSON parsing errors in a block separate from its normal error handling, which skips the logging of stderr and thus loses any error messages from youtube-dl. I've moved the parsing step to before the main error-handling block and now handle parsing errors inside that block, meaning they're now logged alongside youtube-dl's stderr.

In certain cases, such as when using the yt-dlp fork of youtube-dl, an error would result in output of the word "null", which `utils.parse_json` would parse into a `nil`, which was then incorrectly treated as a parsing error, leading to a crash when attempting to concatenate the `nil` error string into a log message: 

```
[ytdl_hook] stack traceback:
[ytdl_hook] 	@ytdl_hook.lua:971: in function 'fn'
[ytdl_hook] 	mp.defaults:579: in function 'handler'
[ytdl_hook] 	mp.defaults:500: in function 'call_event_handlers'
[ytdl_hook] 	mp.defaults:534: in function 'dispatch_events'
[ytdl_hook] 	mp.defaults:493: in function <mp.defaults:492>
[ytdl_hook] 	[C]: at 0x00383b50
[ytdl_hook] 	[C]: at 0x00382560
[ytdl_hook] Lua error: @ytdl_hook.lua:818: attempt to concatenate local 'err' (a nil value)
[ytdl_hook] Exiting...
[cplayer] client removed during hook handling
[...]
[cplayer] Running hook: ytdl_hook/on_load
[cplayer] Sending hook command failed. Removing hook.
[...]
[cplayer] Opening failed or was aborted: https://youtu.be/[redacted]
[cplayer] finished playback, unrecognized file format (reason 4)
[cplayer] Failed to recognize file format.
```

This would render ytdl_hook's `on_load` hook non-functional, preventing playback of any further videos in a playlist. Because the improved error handling checks `parse_json`'s error string to detect parsing errors, this no longer occurs, fixing issues #6285 and #6578.

### CODE REVIEW

**Comment by garoto** [comment]
Impressive detective work. Any example URL(s)?

**Comment by Retusthese** [comment]
If you're using yt-dlp, you can give MPV any blatantly invalid URL that will get passed to ytdl_hook, such as `ytdl://invalid`, and the hook will crash. Note that youtube-dl does not have this problem, as it doesn't output `null` when an error occurs, so you'll get a normal error message about the URL not being valid.

**Comment by garoto** [comment]
Weird, I don't get a `ytdl_hook` crash with my config when using `ytdl://invalid`:

```[ C:\Users\User\AppData\Roaming\mpv ]

> mpv ytdl://invalid
[  0.932660]  ytdl_hook: ERROR: 'invalid' is not a valid URL. Set --default-search "ytsearch" (or run  youtube-dl "ytsearch:invalid" ) to search YouTube
[  0.933313]  ytdl_hook: youtube-dl failed: unexpected error occurred
[  0.934095]     stream: No protocol handler found to open URL ytdl://invalid
[  0.934650]     stream: The protocol is either unsupported, or was disabled at compile-time.
[  0.935090]    cplayer:
[  0.935440]    cplayer: Exiting... (Errors when loading file)

[ C:\Users\User\AppData\Roaming\mpv ]

> rg "ytdl" mpv.conf|rg -v #
ytdl-format="(bestvideo[width<=1920][fps<=60][vcodec^=vp9]+bestaudio[acodec=opus])[protocol!^=http_dash_segments]/best/bestvideo+bestaudio"
ytdl-raw-options=force-ipv4=,ignore-config=,sub-lang="en,pt,en-US,en-GB,pt-BR,pt-PT",write-sub=,cookies="C:\Users\User\cookies_all_20210523121734.txt"
script-opts=ytdl_hook-try_ytdl_first=yes,ytdl_hook-exclude="%.webm$|%.ts$|%.mp3$|%.m3u8$|%.m3u$|%.mkv$|%.mp4$|%.VOB$",autocrop.detect_seconds=2,console-scale=1.5,console-font="Meslo LG M"

```

BUT I do get it with `--no-config`:

```
> mpv --no-config ytdl://invalid
[ytdl_hook]
[ytdl_hook] stack traceback:
[ytdl_hook]     @ytdl_hook.lua:931: in function 'fn'
[ytdl_hook]     mp.defaults:579: in function 'handler'
[ytdl_hook]     mp.defaults:500: in function 'call_event_handlers'
[ytdl_hook]     mp.defaults:534: in function 'dispatch_events'
[ytdl_hook]     mp.defaults:493: in function <mp.defaults:492>
[ytdl_hook]     [C]: at 0x00981e20
[ytdl_hook]     [C]: at 0x00980830
[ytdl_hook] Lua error: @ytdl_hook.lua:789: attempt to concatenate local 'err' (a nil value)
client removed during hook handling
No protocol handler found to open URL ytdl://invalid
The protocol is either unsupported, or was disabled at compile-time.
Sending hook command failed. Removing hook.

Exiting... (Errors when loading file)
```

race stuff?

**Comment by garoto** [comment]
> If you're using yt-dlp [...]

Okie, disregard my post above.

**Comment by Retusthese** [comment]
It *is* odd that you got a crash with regular youtube-dl, though. I tested the latest releases of MPV and youtube-dl with no config files and it didn't happen.

Is your youtube-dl up-to-date? I assume youtube-dl used to output `null`, considering this crash was seen in issues dating back to 2018, which predates any significant forks I'm aware of. Maybe the change that stopped that is fairly recent.

File: player/lua/ytdl_hook.lua [hunk 1]
```diff
@@ -815,7 +815,7 @@ function run_ytdl_hook(url)
     local json, err = utils.parse_json(json)
 
     if (json == nil) then
-        msg.error("failed to parse JSON data: " .. err)
+        msg.error("failed to parse JSON data: " .. (err or "null json"))
```
Comment (avih) [inline player/lua/ytdl_hook.lua:818]: I'd prefer a more correct message, for instance, something like this maybe?
```lua
    if (json == nil) then
        if err then
            msg.error("failed to parse JSON data: " .. err)
        else
            msg.error("null JSON data")
        end
```

Comment (Retusthese) [reply player/lua/ytdl_hook.lua:818]: On second thought, this separate block for json parsing errors shouldn't even exist. I'll move the parsing up so its errors can be handled with the others.

Comment (avih) [reply player/lua/ytdl_hook.lua:818]: Not sure what you mean, show me the code...

**Comment by Retusthese** [comment]
I've moved the json parsing so it happens before the block that handles other youtube-dl errors. Now, parsing errors are handled in the same block, which means you also get youtube-dl's stderr logged when they happen. If the json parses to nil, that's also treated as an error, which avoids the aforementioned crash.

**Comment by avih** [comment]
> I've moved the json parsing...

I don't like this much, because it's a much bigger change, and it's much less obvious that it doesn't regress anything.

I suggest that if you want to enhance the error reporting and/or handling, then you can do that in a separate commit.

I still maintain that your original commit was fine - with the exception that the error message was incorrect, because it didn't fail to parse - it did parse successfully, but the value is null. null might be invalid for whatever reasons, but it's _not_ a failure to parse the output in the specific case which you addressed.

So, IMO bring back the original commit but with the fixed error message, and then, if you want - in a separate commit or PR - do the updated error handling/reporting.

The guiding principle is that every commit does one thing. If you want to fix the exception - do just that. If you want to revise the error handling - do just that, but without implicitly fixing other bugs.

**Comment by Retusthese** [comment]
> I don't like this much, because it's a much bigger change, and it's much less obvious that it doesn't regress anything.

It's a bigger change, but it's mostly moving around the existing error-handling logic to parse earlier. I've tested playable and non-playable videos with youtube-dl and yt-dlp to ensure it works correctly.

> So, IMO bring back the original commit but with the fixed error message, and then, if you want - in a separate commit or PR - do the updated error handling/reporting.

The updated error handling negates the previous commit entirely. If I brought it back, it would be followed by a commit that erases it.

> The guiding principle is that every commit does one thing. If you want to fix the exception - do just that. If you want to revise the error handling - do just that, but without implicitly fixing other bugs.

I want to revise the error handling. This happens to avoid the exception I originally wanted to fix, because it checks for parsing errors in a different way, as it must, since `json` being `nil` is common to all error states. The PR description *does* still focus on the exception, as I wanted to keep the original explanation for context. I'll re-word it to focus on the changes...

**Comment by avih** [comment]
Can you give explicit example cases where the new code is better than the previous code, _other_ than cases which were fixed by the original patch (i.e. where json is `nulll` without parsing errors)?

For each case, please describe exactly what happened before this patch, and what happens after this patch, and how is it better.

**Comment by Retusthese** [comment]
Previously, if youtube-dl encountered an error, printed the error to stderr, and printed text that could not be parsed as JSON to stdout, ytdl_hook would catch the parsing error and only log a message about that error:

```
[ytdl_hook] failed to parse JSON data: error
Failed to recognize file format.

Exiting... (Errors when loading file)
```

After this patch, parsing errors are routed through the same error handler as other errors, which also logs anything youtube-dl printed to stderr. This is better because you can now see errors from youtube-dl that may have resulted in the parsing error. The same command as used above now outputs this:

```
[ytdl_hook] ERROR: Private video
[ytdl_hook] Sign in if you've been granted access to this video
[ytdl_hook] youtube-dl failed: failed to parse JSON data: error
Failed to recognize file format.

Exiting... (Errors when loading file)
```

The outputs above can be reproduced with the following command, which attempts to play a private video and uses an extra youtube-dl option that prints out headers and renders the output unparsable:
`mpv --ytdl-raw-options=print-traffic= https://youtu.be/863SLa6vMT4`
Of course, this command will never output parsable JSON, and so is only useful to demonstrate how parsing errors are handled.

**Comment by avih** [comment]
Thanks.

**Pull request merged**